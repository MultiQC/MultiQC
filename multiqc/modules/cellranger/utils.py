def clean_title_case(col_id):
    title = col_id.title() if col_id[0:1].islower() else col_id
    for str in ["Bc", "bc", "Umi", "Igk", "Igh", "Igl", "Vj", "q30"]:
        title = title.replace(str, str.upper())
    return title


def update_dict(table, headers, rows_list, col_map, colours, namespace):
    """update the data dict and headers dict"""

    for col_name, col_data in rows_list:
        if col_name in col_map:
            # Sanitize numeric data
            is_percentage = "%" in col_data
            col_data = col_data.replace(",", "").replace("%", "")

            # Convert to float when possible
            try:
                col_data = float(col_data)
            except ValueError:
                col_data = col_data

            col_id = col_map[col_name]
            table[col_id] = col_data
            headers[col_id] = {
                "rid": "{}_{}".format(namespace, col_id.replace(" ", "_").replace("/", "_")).lower(),
                "title": clean_title_case(col_id),
                "description": col_name,
                "namespace": namespace,
                "scale": colours.get(col_id, "RdYlGn" if is_percentage else "GnBu"),
            }
            if is_percentage:
                headers[col_id].update(
                    {
                        "suffix": "%",
                        "max": 100,
                        "min": 0,
                    }
                )

            # Assign shared keys
            if col_id == "estimated cells":
                headers[col_id]["shared_key"] = "cell_count"
                headers[col_id]["format"] = "{:,.0f}"
            if col_id.endswith("reads/cell"):
                headers[col_id]["shared_key"] = "reads_per_cell"
                headers[col_id]["format"] = "{:,.0f}"

    return table, headers


def set_hidden_cols(headers, col_names):
    """Set the hidden columns"""

    for col_name in col_names:
        try:
            headers[col_name]["hidden"] = True
        except KeyError:
            pass

    return headers


def parse_bcknee_data(data, s_name, max_idx=1000):
    """parse data for bc knee plot from cellranger dict"""

    value_dict = dict()
    for idx, data_series in enumerate(data):
        if idx > max_idx:
            break
        if len(data_series["x"]) == 0:
            continue
        id = f"{s_name}_{data_series['name']}"
        if id not in value_dict.keys():
            value_dict[id] = dict()
        value_dict[id].update(transform_data(data_series))

    return value_dict


def transform_data(data):
    """Transform x:list,y:list data to a dict of x_val:y_val"""

    value_dict = dict()
    for idx, row in enumerate(data["x"]):
        if row > 0 and data["y"][idx] > 0:
            value_dict[row] = data["y"][idx]

    return value_dict
