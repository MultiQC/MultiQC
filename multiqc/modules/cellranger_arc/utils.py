import logging
from typing import Dict, List, Tuple

from multiqc import config

log = logging.getLogger(__name__)


def table_data_and_headers(
    rows_list: List[Tuple[str, str]],
    help: List[Tuple[str, List[str]]],
) -> Tuple[Dict, Dict]:
    """Update the data dict and headers dict"""
    table = dict()
    headers = dict()

    help_dict = {lst[0]: lst[1][0].strip(".") for lst in help}
    for col_name, col_data in rows_list:
        # Sanitize numeric data
        is_percentage = "%" in col_data
        col_data = col_data.replace(",", "").replace("%", "")
        is_integer = col_data.isdigit()

        # Convert to float when possible
        try:
            col_data = float(col_data)
        except ValueError:
            col_data = col_data

        table[col_name] = col_data

        # Assign shared/regular keys
        if col_name == "Sequenced read pairs":
            headers[col_name] = {
                "title": col_name,
                "description": f"{help_dict[col_name]} ({config.read_count_desc})",
                "modify": lambda x: x * config.read_count_multiplier,
                "shared_key": "read_count",
                "format": "{:,.0f}",
            }
        else:
            headers[col_name] = {
                "title": col_name,
                "description": help_dict[col_name],
            }
        if col_name == "Estimated number of cells":
            headers[col_name]["shared_key"] = "cell_count"
            headers[col_name]["title"] = "Est. cells"
            headers[col_name]["description"] += " (which estimates the number of cells)"

        if col_name == "Fraction of high-quality fragments in cells":
            headers[col_name]["title"] = "High-qual fragments"

        if is_integer:
            headers[col_name]["format"] = "{:,.0f}"

        if is_percentage:
            headers[col_name].update(
                {
                    "suffix": "%",
                    "max": 100,
                    "min": 0,
                }
            )

    return table, headers


def set_hidden_cols(headers, col_names):
    """Set the hidden columns"""

    for col_name in col_names:
        try:
            headers[col_name]["hidden"] = True
        except KeyError:
            pass

    return headers


def subset_header(data, cols, namespace=None):
    """Subsets the headers to only columns in cols. Adds colour and namespace"""

    headers = dict()
    for key, val in cols.items():
        headers[key] = data[key]
        headers[key]["scale"] = val
        headers[key]["namespace"] = namespace

    return headers


def extract_plot_data(data):
    """Extracts plot data"""

    plot_data = dict()
    axes_data = data["data"][0]
    plot_data = dict(zip(axes_data["x"], axes_data["y"]))
    return plot_data
