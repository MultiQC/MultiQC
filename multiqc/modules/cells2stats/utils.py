def find_entry(json_list, key, value, default_value):
    """
    Find an entry in a list of dictionaries by key and value
    """
    for entry in json_list:
        if entry.get(key, "") == value:
            return entry
    return default_value


def json_decode_float(value):
    """
    Decode a JSON float value, converting it to NaN if it is -999
    """
    value = float(value)
    if abs(value + 999) < 0.00001:
        return float("nan")
    return value


def json_decode_int(value):
    """
    Decode a JSON int value, converting it to 0 if it is -999
    """
    value = int(value)
    if abs(value + 999) < 0.00001:
        return 0
    return value


def is_nan(value):
    """
    Check if a value is NaN
    """
    return value != value


def summarize_batch_names(c2s_run_data):
    """
    Generate a list of barcoding batch names from the cells2stats report
    """
    batch_names = set()
    for run_name in c2s_run_data:
        run_data = c2s_run_data[run_name]
        for well_data in run_data.get("CytoStats", {}).get("Wells", []):
            for batch_data in well_data.get("Batches", []):
                batch_name = batch_data.get("BatchName", "")
                if batch_name != "" and not batch_name.startswith("CP"):
                    batch_names.add(batch_data["BatchName"])

    return sorted(batch_names)
