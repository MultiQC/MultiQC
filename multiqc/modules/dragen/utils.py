from collections import OrderedDict
from multiqc import config


read_format = '{:,.1f}'
if config.read_count_multiplier == 1:
    read_format = '{:,.0f}'
# read_format += '&nbsp;' + config.read_count_prefix

base_format = '{:,.1f}&nbsp;'
if config.base_count_multiplier == 1:
    base_format = '{:,.0f}'
elif config.base_count_multiplier == 0.000000001:
    base_format = '{:,.2f}'
# base_format += '&nbsp;' + config.base_count_prefix


class Metric:
    def __init__(self, id, title, in_genstats=None, in_own_tabl=None, unit=None,
                 descr=None, fmt=None, modify=None, namespace=None, precision=None,
                 the_higher_the_worse=False):
        self.id = id
        self.title = title
        self.in_genstats = in_genstats
        self.in_own_tabl = in_own_tabl
        self.unit = unit
        self.descr = descr
        self.fmt = fmt
        self.modify = modify
        self.namespace = namespace
        self.precision = precision
        self.the_higher_the_worse = the_higher_the_worse


def make_headers(parsed_metric_ids, metrics):
    # Init general stats table
    genstats_headers = OrderedDict()

    # Init headers for an own separate table
    own_tabl_headers = OrderedDict()

    for metric in metrics:
        col = dict(
            title=metric.title,
            description=metric.descr,
            min=0,
        )

        # choosing color based on metric namespace, guessing namespace by unit
        if metric.unit == 'reads':
            col['scale'] = 'Greens'
        elif metric.unit == 'bases':
            col['scale'] = 'Greys'
        elif metric.unit == 'bp':
            col['scale'] = 'Purples'
        elif metric.unit == 'x' or metric.namespace == 'Coverage':
            col['scale'] = 'Blues'
            if metric.unit == '%':
                col['scale'] = 'RdBu'
                if metric.the_higher_the_worse:
                    col['scale'] = 'YlOrRd'
        elif metric.namespace == 'Variants':
            col['scale'] = 'Purples'

        if metric.id + ' pct' in parsed_metric_ids:
            # if % value is available, showing it instead of the number value; the number value will be hidden
            pct_col = dict(
                col,
                description=metric.descr.replace(', {}', '').replace('Number of ', '% of '),
                max=100,
                suffix='%',
            )
            if metric.unit == 'reads':
                pct_col['scale'] = 'RdYlGn'
            elif metric.unit == 'bases':
                pct_col['scale'] = 'RdGy'
            elif metric.unit == 'x' or metric.namespace == 'Coverage':
                pct_col['scale'] = 'RdBu'
            if metric.the_higher_the_worse:
                pct_col['scale'] = 'YlOrRd'

            if metric.precision is not None:
                pct_col['format'] = '{:,.' + str(metric.precision) + 'f}'

            if metric.in_own_tabl is not None:
                show = metric.in_own_tabl == '%'
                own_tabl_headers[metric.id + ' pct'] = dict(pct_col, hidden=not show)
            if metric.in_genstats is not None and metric.in_own_tabl != '#':
                genstats_headers[metric.id + ' pct'] = dict(pct_col, hidden=metric.in_genstats == 'hid')

        if metric.unit == 'reads':
            col['description'] = col['description'].format(config.read_count_desc)
            col['title'] = config.read_count_prefix + ' ' + col['title']
            col['modify'] = lambda x: x * config.read_count_multiplier
            col['shared_key'] = 'read_count'
            col['format'] = read_format
        elif metric.unit == 'bases':
            col['description'] = col['description'].format(config.base_count_desc)
            col['title'] = config.base_count_prefix + ' ' + col['title']
            col['modify'] = lambda x: x * config.base_count_multiplier
            col['shared_key'] = 'base_count'
            col['format'] = base_format
        elif metric.unit == 'len':
            col['suffix'] = ' bp'
            col['format'] = '{:,.0f}'
        elif metric.unit == 'x':
            col['suffix'] = ' x'
            col['format'] = '{:,.1f}'
        elif metric.unit == '%':
            col['suffix'] = ' %'
            col['format'] = '{:,.1f}'
        elif any(metric.descr.startswith(pref) for pref in ('Total number of ', 'The number of ', 'Number of ')):
            col['format'] = '{:,.0f}'
        if metric.precision is not None:
            col['format'] = '{:,.' + str(metric.precision) + 'f}'

        if metric.in_own_tabl is not None:
            own_tabl_headers[metric.id] = dict(col, hidden=metric.in_own_tabl != '#')
        if metric.in_genstats is not None:
            if metric.in_genstats == '#':
                genstats_headers[metric.id] = dict(col)
            elif metric.in_genstats == 'hid' and metric.in_own_tabl == '#':
                genstats_headers[metric.id] = dict(col, hidden=metric.in_genstats == 'hid')

    return genstats_headers, own_tabl_headers


def exist_and_number(data, *metrics):
    return all(isinstance(data.get(m, None), int) or isinstance(data.get(m, None), float) for m in metrics)
