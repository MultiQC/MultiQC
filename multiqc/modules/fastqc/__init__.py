from .fastqc import MultiqcModule

__all__ = ["MultiqcModule"]
