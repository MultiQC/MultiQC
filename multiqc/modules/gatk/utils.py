def parse_report(lines, table_names):
    """Parse a GATK report https://software.broadinstitute.org/gatk/documentation/article.php?id=1244

    Only GATTable entries are parsed.  Tables are returned as a dict of tables.
    Each table is a dict of arrays, where names correspond to column names, and arrays
    correspond to column values.

    Args:
        lines (file handle): an iterable over the lines of a GATK report.
        table_names (dict): a dict with keys that are GATK report table names
            (e.g. "#:GATKTable:Quantized:Quality quantization map"), and values that are the
            keys in the returned dict.

    Returns:
        {
            table_1:
                {
                    col_1: [ val_1, val_2, ... ]
                    col_2: [ val_1, val_2, ... ]
                    ...
                }
            table_2:
                ...
        }
    """

    report = dict()
    lines = (line for line in lines)
    for line in lines:
        line = line.rstrip()
        if line in table_names.keys():
            report[table_names[line]] = parse_gatk_report_table(lines)
    return report


def parse_gatk_report_table(lines):
    headers = next(lines).rstrip().split()
    table = {h: [] for h in headers}
    for line in lines:
        line = line.rstrip()

        # testing to see if we have reached the end of a table in a GATKReport
        if line == "":
            break

        for index, value in enumerate(line.split()):
            table[headers[index]].append(value)
    return table
