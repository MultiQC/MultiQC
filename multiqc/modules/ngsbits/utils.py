from typing import Dict, Union, Tuple

import xml.etree.ElementTree
import re


def parse_qcml_by(qcml_contents, tag) -> Tuple[Dict[str, Union[float, str]], Dict[str, Tuple[str, str]]]:
    """Parse a qcML file and return key-value pairs from the quality parameter entries."""
    root = xml.etree.ElementTree.fromstring(qcml_contents)
    values: Dict[str, Union[float, str]] = dict()
    params: Dict[str, Tuple[str, str]] = dict()

    for qp in root.findall(".//{http://www.prime-xs.eu/ms/qcml}%s" % tag):
        # skip n/a values
        if qp.attrib["value"].startswith("n/a"):
            continue

        # replace 'percentage' with '%'
        qp_name = re.sub(r" percentage$", " %", qp.attrib["name"])

        try:
            values[qp_name] = float(qp.attrib["value"])
        except ValueError:
            values[qp_name] = qp.attrib["value"]

        # add description and accession number of the parameter to the header
        params[qp_name] = (qp.attrib["description"], qp.attrib["accession"])
    return values, params
