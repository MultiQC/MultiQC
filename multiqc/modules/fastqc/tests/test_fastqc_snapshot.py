import pytest
from multiqc.utils import testing
from multiqc.utils.testing import BaseModuleTest
from multiqc.modules.fastqc.fastqc import MultiqcModule

@pytest.fixture
def data_dir():
    return testing.data_dir()

@pytest.fixture
def snapshot_config():
    """Configuration for snapshot testing."""
    return {
        "preserve_module_raw_data": True,
        "strict": True,
        "make_data_dir": False,
    }

class TestFastqcModule(BaseModuleTest):
    MODULE_CLASS = MultiqcModule
    MODULE_NAME = "fastqc"

    @pytest.mark.parametrize("subdir", [
        "",  # root fastqc_data.txt
        "v0.11.2",  # zipped report
        "zero_reads",  # zero reads
        "nan_reads",  # nan reads
        "groups",  # group/multiple samples
        "issue_1941_unicode",  # unicode sample names
        "issue_2343_numerical_samples",  # numerical sample names
    ])
    def test_fastqc_snapshot(self, data_dir, snapshot, snapshot_config, subdir):
        module_data_dir = data_dir / "modules" / self.MODULE_NAME / subdir
        # Find all fastqc_data.txt and *_fastqc.zip in the directory
        data_files = list(module_data_dir.glob("fastqc_data.txt")) + list(module_data_dir.glob("*_fastqc.zip"))
        if not data_files:
            pytest.skip(f"No fastqc test files found in {module_data_dir}")
        
        # Run the module test on the specific subdirectory
        module_snapshot = testing.run_module_test(
            module_class=self.MODULE_CLASS,
            data_files=data_files,
            config_updates=snapshot_config,
        )
        
        # Assert basic data integrity
        self.assert_module_data_integrity(module_snapshot)
        
        # Snapshot the raw data output from the module
        raw_data = module_snapshot.get_saved_raw_data()
        assert raw_data == snapshot 