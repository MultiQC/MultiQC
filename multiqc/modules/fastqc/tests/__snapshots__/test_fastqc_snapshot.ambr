# serializer version: 1
# name: TestFastqcModule.test_fastqc_snapshot[]
  dict({
    'multiqc_fastqc': dict({
      'SRR1067503_1': dict({
        '%GC': 44.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'SRR1067503_1.fastq.gz',
        'Sequence length': 36.0,
        'Sequences flagged as poor quality': 0.0,
        'Total Sequences': 30680980.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 36.0,
        'basic_statistics': 'pass',
        'kmer_content': 'fail',
        'median_sequence_length': 36,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'warn',
        'per_base_sequence_quality': 'pass',
        'per_sequence_gc_content': 'warn',
        'per_sequence_quality_scores': 'pass',
        'per_tile_sequence_quality': 'pass',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'pass',
        'total_deduplicated_percentage': 86.841642,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[groups]
  dict({
    'multiqc_fastqc': dict({
      'single': dict({
        '%GC': 50.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'single.fastq.gz',
        'Sequence length': '91-108',
        'Sequences flagged as poor quality': 0.0,
        'Total Bases': '26.6 Mbp',
        'Total Sequences': 266736.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 100.052051,
        'basic_statistics': 'pass',
        'median_sequence_length': 100,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'fail',
        'per_base_sequence_quality': 'fail',
        'per_sequence_gc_content': 'fail',
        'per_sequence_quality_scores': 'fail',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'warn',
        'total_deduplicated_percentage': 98.227962,
      }),
      'test_1': dict({
        '%GC': 50.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'test_1.fastq.gz',
        'Sequence length': '91-108',
        'Sequences flagged as poor quality': 0.0,
        'Total Bases': '26.6 Mbp',
        'Total Sequences': 266736.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 100.052051,
        'basic_statistics': 'pass',
        'median_sequence_length': 100,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'fail',
        'per_base_sequence_quality': 'fail',
        'per_sequence_gc_content': 'fail',
        'per_sequence_quality_scores': 'fail',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'warn',
        'total_deduplicated_percentage': 98.227962,
      }),
      'test_2': dict({
        '%GC': 48.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'test_2.fastq.gz',
        'Sequence length': '90-107',
        'Sequences flagged as poor quality': 0.0,
        'Total Bases': '26.6 Mbp',
        'Total Sequences': 266736.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 100.050916,
        'basic_statistics': 'pass',
        'median_sequence_length': 100,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'warn',
        'per_base_sequence_quality': 'fail',
        'per_sequence_gc_content': 'warn',
        'per_sequence_quality_scores': 'fail',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'warn',
        'total_deduplicated_percentage': 99.006463,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[issue_1941_unicode]
  dict({
    'multiqc_fastqc': dict({
      'run1_2': dict({
        '%GC': 48.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'run1_2.fq',
        'Sequence length': '150-154',
        'Sequences flagged as poor quality': 0.0,
        'Total Bases': '4.5 kbp',
        'Total Sequences': 30.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 150.7,
        'basic_statistics': 'pass',
        'median_sequence_length': 150,
        'overrepresented_sequences': 'fail',
        'per_base_n_content': 'fail',
        'per_base_sequence_content': 'fail',
        'per_base_sequence_quality': 'pass',
        'per_sequence_gc_content': 'fail',
        'per_sequence_quality_scores': 'fail',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'warn',
        'total_deduplicated_percentage': 93.333333,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[issue_2343_numerical_samples]
  dict({
    'multiqc_fastqc': dict({
      '00050101': dict({
        '%GC': 41.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': '00050101.aligned-bwa.sorted-samtools.merged.rh.dedupped.bam',
        'Sequence length': '30-151',
        'Sequences flagged as poor quality': 0.0,
        'Total Sequences': 873870503.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 148.774488,
        'basic_statistics': 'pass',
        'median_sequence_length': 151,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'fail',
        'per_base_sequence_quality': 'pass',
        'per_sequence_gc_content': 'warn',
        'per_sequence_quality_scores': 'pass',
        'per_tile_sequence_quality': 'pass',
        'sequence_duplication_levels': 'warn',
        'sequence_length_distribution': 'warn',
        'total_deduplicated_percentage': 54.17663,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[nan_reads]
  dict({
    'multiqc_fastqc': dict({
      'sample1_S1_L001_R2_001': dict({
        '%GC': 44.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'sample1_S1_L001_R2_001.fastq.gz',
        'Sequence length': 301.0,
        'Sequences flagged as poor quality': 0.0,
        'Total Sequences': 82330.0,
        'adapter_content': 'fail',
        'avg_sequence_length': 301.0,
        'basic_statistics': 'pass',
        'median_sequence_length': 301,
        'overrepresented_sequences': 'warn',
        'per_base_n_content': 'fail',
        'per_base_sequence_content': 'fail',
        'per_base_sequence_quality': 'fail',
        'per_sequence_gc_content': 'pass',
        'per_sequence_quality_scores': 'warn',
        'per_tile_sequence_quality': 'fail',
        'sequence_duplication_levels': 'fail',
        'sequence_length_distribution': 'pass',
        'total_deduplicated_percentage': 29.744929,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[v0.11.2]
  dict({
    'multiqc_fastqc': dict({
      'SRR1067505_1': dict({
        '%GC': 47.0,
        'Encoding': 'Sanger / Illumina 1.9',
        'File type': 'Conventional base calls',
        'Filename': 'SRR1067505_1.fastq.gz',
        'Sequence length': 36.0,
        'Sequences flagged as poor quality': 0.0,
        'Total Sequences': 18361776.0,
        'adapter_content': 'pass',
        'avg_sequence_length': 36.0,
        'basic_statistics': 'pass',
        'kmer_content': 'fail',
        'median_sequence_length': 36,
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'pass',
        'per_base_sequence_quality': 'pass',
        'per_sequence_gc_content': 'pass',
        'per_sequence_quality_scores': 'pass',
        'per_tile_sequence_quality': 'fail',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'pass',
        'total_deduplicated_percentage': 91.788429,
      }),
    }),
  })
# ---
# name: TestFastqcModule.test_fastqc_snapshot[zero_reads]
  dict({
    'multiqc_fastqc': dict({
      'SK-GBD-000919.1': dict({
        '%GC': 0.0,
        'Encoding': 'Illumina 1.5',
        'File type': 'null',
        'Filename': 'SK-GBD-000919.1.fastq',
        'Sequence length': 0.0,
        'Sequences flagged as poor quality': 0.0,
        'Total Sequences': 0.0,
        'adapter_content': 'warn',
        'basic_statistics': 'pass',
        'overrepresented_sequences': 'pass',
        'per_base_n_content': 'pass',
        'per_base_sequence_content': 'pass',
        'per_sequence_gc_content': 'pass',
        'sequence_duplication_levels': 'pass',
        'sequence_length_distribution': 'pass',
        'total_deduplicated_percentage': 100.0,
      }),
    }),
  })
# ---
